File: stream_fetcher.js
#!/usr/bin/env node
/**
 * Vidstorm Stream Fetcher - Pure JavaScript
 * Standalone script to fetch streams using headless browser
 */

const axios = require('axios');
const { chromium } = require('playwright');

const TMDB_API_KEY = "54e00466a09676df57ba51c4ca30b1a6";

/**
 * Fetch TMDB metadata
 */
async function getTmdbMetadata(tmdbId) {
  try {
    const url = `https://api.themoviedb.org/3/movie/${tmdbId}?api_key=${TMDB_API_KEY}`;
    const response = await axios.get(url, { timeout: 10000 });
    return response.data;
  } catch (error) {
    console.error('TMDB error:', error.message);
    return null;
  }
}

/**
 * Extract token using headless browser
 */
async function extractToken(tmdbId) {
  console.log(`[Browser] Starting extraction for movie ${tmdbId}...`);
  
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({
    userAgent: 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36'
  });
  const page = await context.newPage();
  
  let capturedToken = null;
  
  // Intercept network requests
  await page.route('**/*', (route, request) => {
    const url = request.url();
    const match = url.match(/\/api\/(?:movie|tv)\/([A-Za-z0-9_-]{20,40})/);
    if (match && !capturedToken) {
      capturedToken = match[1];
      console.log(`[Browser] ✓ Captured token: ${capturedToken}`);
    }
    route.continue();
  });
  
  // Load movie page
  const movieUrl = `https://vidstorm.ru/movie/${tmdbId}`;
  console.log(`[Browser] Loading ${movieUrl}...`);
  
  try {
    await page.goto(movieUrl, { waitUntil: 'networkidle', timeout: 30000 });
  } catch (e) {
    // Expected timeout
  }
  
  // Wait for API calls
  await page.waitForTimeout(5000);
  
  // Try page content as fallback
  if (!capturedToken) {
    try {
      const content = await page.content();
      const match = content.match(/\/api\/movie\/([A-Za-z0-9_-]{20,40})/);
      if (match) {
        capturedToken = match[1];
        console.log(`[Browser] ✓ Extracted from HTML: ${capturedToken}`);
      }
    } catch (e) {}
  }
  
  await browser.close();
  return capturedToken;
}

/**
 * Fetch streams for a movie
 */
async function fetchStreams(tmdbId, providedToken = null) {
  const result = {
    movie_id: tmdbId,
    token: null,
    sources: [],
    subtitles: [],
    metadata: null
  };
  
  // Get TMDB metadata
  console.log('[TMDB] Fetching metadata...');
  const metadata = await getTmdbMetadata(tmdbId);
  if (metadata) {
    result.metadata = {
      title: metadata.title,
      overview: metadata.overview,
      poster: metadata.poster_path ? `https://image.tmdb.org/t/p/w500${metadata.poster_path}` : null,
      release_date: metadata.release_date,
      runtime: metadata.runtime,
      vote_average: metadata.vote_average
    };
    console.log(`[TMDB] ✓ ${metadata.title} (${metadata.release_date})`);
  }
  
  // Get token
  let token = providedToken;
  if (!token) {
    token = await extractToken(tmdbId);
  }
  
  if (!token) {
    result.error = 'Failed to extract token';
    return result;
  }
  
  result.token = token;
  
  // Fetch streams
  const apiUrl = `https://vidstorm.ru/api/movie/${token}`;
  console.log(`[API] Fetching streams...`);
  
  try {
    const response = await axios.get(apiUrl, {
      headers: {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36',
        'Accept': '*/*',
        'Referer': `https://vidstorm.ru/movie/${tmdbId}`,
        'Origin': 'https://vidstorm.ru'
      },
      timeout: 15000
    });
    
    if (response.status === 200) {
      const data = response.data;
      console.log(`[API] ✓ Found ${Object.keys(data).length} sources`);
      
      for (const [name, source] of Object.entries(data)) {
        if (source && source.url) {
          result.sources.push({
            name,
            url: source.url,
            type: source.type || 'hls',
            language: source.language,
            flag: source.flag
          });
        }
      }
    }
  } catch (error) {
    result.error = `API error: ${error.message}`;
  }
  
  // Fetch subtitles
  try {
    const subUrl = `https://sub.vdrk.site/v2/movie/${tmdbId}`;
    const response = await axios.get(subUrl, { timeout: 10000 });
    if (response.status === 200) {
      result.subtitles = response.data;
      console.log(`[Subtitles] ✓ Found ${result.subtitles.length} subtitles`);
    }
  } catch (e) {
    console.log('[Subtitles] ✗ None found');
  }
  
  return result;
}

/**
 * Main CLI
 */
async function main() {
  const args = process.argv.slice(2);
  
  if (args.length < 1) {
    console.log('Usage: node stream_fetcher.js <tmdb_id> [token]');
    console.log('Examples:');
    console.log('  node stream_fetcher.js 557');
    console.log('  node stream_fetcher.js 557 TWkB6xPDtW-pyk5ryfskpQ');
    process.exit(1);
  }
  
  const tmdbId = args[0];
  const token = args[1] || null;
  
  console.log('='.repeat(60));
  console.log(`Fetching streams for TMDB ID: ${tmdbId}`);
  console.log('='.repeat(60));
  console.log('');
  
  const result = await fetchStreams(tmdbId, token);
  
  console.log('');
  console.log('='.repeat(60));
  console.log('RESULT:');
  console.log('='.repeat(60));
  console.log(JSON.stringify(result, null, 2));
  
  if (result.sources.length > 0) {
    console.log('');
    console.log('✅ STREAM SOURCES:');
    result.sources.forEach((src, i) => {
      console.log(`  ${i + 1}. ${src.name} (${src.language || 'unknown'})`);
      console.log(`     URL: ${src.url.substring(0, 70)}...`);
    });
  } else {
    console.log('');
    console.log('❌ No stream sources found');
    if (result.error) {
      console.log(`Error: ${result.error}`);
    }
  }
}

main().catch(console.error);

File: stream_fetcher.py
#!/usr/bin/env python3
"""
Vidstorm Stream Fetcher - Uses headless browser to extract working tokens
This is the ONLY reliable way to get streams since server-side tokens get 403
"""

import json
import re
import requests
from urllib.parse import urljoin

# Try to import playwright
try:
    from playwright.sync_api import sync_playwright
    PLAYWRIGHT_AVAILABLE = True
except ImportError:
    PLAYWRIGHT_AVAILABLE = False
    print("Playwright not installed. Run: pip install playwright")
    print("Then: playwright install chromium")


class VidstormStreamFetcher:
    """Fetches vidstorm streams using headless browser for token extraction"""
    
    TMDB_API_KEY = "54e00466a09676df57ba51c4ca30b1a6"
    
    def __init__(self):
        self.session = requests.Session()
        self.session.headers.update({
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
        })
    
    def get_tmdb_metadata(self, tmdb_id: str):
        """Fetch movie metadata from TMDB"""
        try:
            url = f"https://api.themoviedb.org/3/movie/{tmdb_id}?api_key={self.TMDB_API_KEY}"
            resp = self.session.get(url, timeout=10)
            if resp.status_code == 200:
                return resp.json()
        except Exception as e:
            print(f"TMDB error: {e}")
        return None
    
    def extract_token_with_browser(self, tmdb_id: str):
        """
        Use headless browser to extract working token from vidstorm
        This is the ONLY reliable method - server-generated tokens get 403
        """
        if not PLAYWRIGHT_AVAILABLE:
            print("ERROR: Playwright not available. Install it first.")
            return None
        
        try:
            with sync_playwright() as p:
                # Launch browser
                browser = p.chromium.launch(headless=True)
                context = browser.new_context(
                    user_agent="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36"
                )
                page = context.new_page()
                
                # Store captured token
                captured_token = None
                
                # Intercept network requests to capture the token
                def handle_route(route, request):
                    nonlocal captured_token
                    url = request.url
                    
                    # Look for /api/movie/ or /api/tv/ requests
                    match = re.search(r'/api/(?:movie|tv)/([A-Za-z0-9_-]{20,40})', url)
                    if match and not captured_token:
                        captured_token = match.group(1)
                        print(f"[Browser] Captured token: {captured_token}")
                    
                    route.continue_()
                
                page.route("**/*", handle_route)
                
                # Navigate to movie page
                movie_url = f"https://vidstorm.ru/movie/{tmdb_id}"
                print(f"[Browser] Loading {movie_url}...")
                
                try:
                    page.goto(movie_url, wait_until="networkidle", timeout=30000)
                except Exception as e:
                    print(f"[Browser] Page load timeout (expected): {e}")
                
                # Wait a bit for API calls
                page.wait_for_timeout(5000)
                
                # Also try to extract from page content
                if not captured_token:
                    try:
                        content = page.content()
                        match = re.search(r'/api/movie/([A-Za-z0-9_-]{20,40})', content)
                        if match:
                            captured_token = match.group(1)
                            print(f"[Browser] Extracted token from HTML: {captured_token}")
                    except:
                        pass
                
                browser.close()
                return captured_token
                
        except Exception as e:
            print(f"[Browser] Error: {e}")
            return None
    
    def fetch_streams(self, tmdb_id: str, token: str = None):
        """
        Fetch streams for a movie
        If no token provided, uses browser to extract one
        """
        result = {
            "movie_id": tmdb_id,
            "token": token,
            "sources": [],
            "subtitles": [],
            "metadata": None
        }
        
        # Get TMDB metadata
        metadata = self.get_tmdb_metadata(tmdb_id)
        if metadata:
            result["metadata"] = {
                "title": metadata.get("title"),
                "overview": metadata.get("overview"),
                "poster": f"https://image.tmdb.org/t/p/w500{metadata.get('poster_path')}" if metadata.get('poster_path') else None,
                "release_date": metadata.get("release_date"),
                "runtime": metadata.get("runtime")
            }
        
        # If no token, extract with browser
        if not token:
            print(f"[Stream Fetcher] No token provided, extracting with browser...")
            token = self.extract_token_with_browser(tmdb_id)
            if token:
                result["token"] = token
                result["token_source"] = "browser"
            else:
                result["error"] = "Failed to extract token with browser"
                return result
        else:
            result["token_source"] = "provided"
        
        # Fetch streams with the token
        api_url = f"https://vidstorm.ru/api/movie/{token}"
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36",
            "Accept": "*/*",
            "Referer": f"https://vidstorm.ru/movie/{tmdb_id}",
            "Origin": "https://vidstorm.ru"
        }
        
        try:
            print(f"[Stream Fetcher] Fetching streams from API...")
            resp = self.session.get(api_url, headers=headers, timeout=15)
            print(f"[Stream Fetcher] API status: {resp.status_code}")
            
            if resp.status_code == 200:
                data = resp.json()
                print(f"[Stream Fetcher] Got {len(data)} sources")
                
                for name, source in data.items():
                    if isinstance(source, dict) and source.get('url'):
                        result["sources"].append({
                            "name": name,
                            "url": source['url'],
                            "type": source.get('type', 'hls'),
                            "language": source.get('language'),
                            "quality": self._extract_quality(source['url'])
                        })
            else:
                result["error"] = f"API returned {resp.status_code}"
                
        except Exception as e:
            result["error"] = str(e)
        
        # Fetch subtitles
        try:
            sub_url = f"https://sub.vdrk.site/v2/movie/{tmdb_id}"
            resp = self.session.get(sub_url, timeout=10)
            if resp.status_code == 200:
                result["subtitles"] = resp.json()
        except:
            pass
        
        return result
    
    def _extract_quality(self, url: str):
        """Try to extract quality from stream URL"""
        patterns = [
            r'(\d+)p',
            r'(?:^|/)(\d+)[xX]',
            r'(?:^|[_-])(\d+)[xX]'
        ]
        for pattern in patterns:
            match = re.search(pattern, url)
            if match:
                return match.group(1) + "p"
        return "unknown"


def main():
    """CLI interface"""
    import sys
    
    if len(sys.argv) < 2:
        print("Usage: python stream_fetcher.py <tmdb_id> [token]")
        print("Examples:")
        print("  python stream_fetcher.py 557")
        print("  python stream_fetcher.py 557 TWkB6xPDtW-pyk5ryfskpQ")
        return
    
    tmdb_id = sys.argv[1]
    token = sys.argv[2] if len(sys.argv) > 2 else None
    
    fetcher = VidstormStreamFetcher()
    result = fetcher.fetch_streams(tmdb_id, token)
    
    print("\n" + "="*60)
    print("RESULT:")
    print("="*60)
    print(json.dumps(result, indent=2))
    
    if result.get("sources"):
        print(f"\n✅ Found {len(result['sources'])} stream sources!")
        for src in result["sources"]:
            print(f"  - {src['name']}: {src['quality']} - {src['url'][:60]}...")
    else:
        print("\n❌ No stream sources found")
        if result.get("error"):
            print(f"Error: {result['error']}")


if __name__ == "__main__":
    main()
